### PR TITLE
Add a provider for Cohere

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add Cohere as a gateway provider.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/gateway/providers/test_cohere.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                    2.65625,
+                    -0.30126953,
+                    -2.3554688,
+                    1.2597656,
+                ],
+            ],
+            "metadata": {
+                "model": "embed-english-light-v2.0",
+                "route_type": "llm/v1/embeddings",
+            },
```
Comment (BenWilson2) [inline tests/gateway/providers/test_cohere.py:128]:
```suggestion
            "metadata": {
                "model": "embed-english-light-v2.0",
                "route_type": "llm/v1/embeddings",
                "input_tokens": None,
                "output_tokens": None,
                "total_tokens": None,
            },
```

File: tests/gateway/providers/test_cohere.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        payload = {
+            "prompt": "This is a test",
+        }
+        response = await provider.completions(completions.RequestPayload(**payload))
+        assert jsonable_encoder(response) == {
+            "candidates": [{"text": "This is a tes!", "metadata": {}}],
+            "metadata": {
+                "model": "command",
+                "route_type": "llm/v1/completions",
+            },
```
Comment (BenWilson2) [inline tests/gateway/providers/test_cohere.py:51]:
```suggestion
            "metadata": {
                "model": "command",
                "route_type": "llm/v1/completions",
                "input_tokens": None,
                "output_tokens": None,
                "total_tokens": None,
            },
```

**Comment by BenWilson2** [comment]
non-blocking question: Do you think that when we add a provider we should have a pseudo-integration test added to fluent or client API where we start the Gateway server with at least 1 route from the provider so that we can validate that the full instantiation / creation of the provider dynamic routes work (while mocking the external provider request call)?

**Review by BenWilson2 — APPROVED** [review]
LGTM with a small question on testing

File: tests/gateway/providers/test_cohere.py (hunk 1 above)
Comment (harupy) [reply tests/gateway/providers/test_cohere.py:128]: I thought these were optional.

**Comment by harupy** [comment]
Given all the providers implement the same abstract class, I think we don't need to test all the providers. We can just pick one and use it. If it works for openai, it should work for other providers.

**Pull request merged**